cranelift/frontend/src/frontend.rs

```diff
@@ -382,21 +397,56 @@ impl<'a> FunctionBuilder<'a> {
         self.handle_ssa_side_effects(side_effects);
     }
 
-    /// Declares the type of a variable, so that it can be used later (by calling
-    /// [`FunctionBuilder::use_var`]). This function will return an error if the variable
-    /// has been previously declared.
-    pub fn try_declare_var(&mut self, var: Variable, ty: Type) -> Result<(), DeclareVariableError> {
+    /// Declares the type of a variable.
+    ///
+    /// This allows the variable to be used later (by calling
+    /// [`FunctionBuilder::use_var`]).
+    ///
+    /// If `stack_map_behavior` is `StackMapBehavior::Include`, then all values
+    /// yielded by using the resulting variable will automatically be declared
+    /// as needing to be present in stack maps. See also
+    /// [`FunctionBuilder::declare_value_needs_stack_map`].
+    ///
+    /// # Errors
+    ///
+    /// This function will return an error if the variable has been previously
+    /// declared.
+    ///
+    /// # Panics
+    ///
+    /// Panics if `needs_stack_map` is `true` and the given type is larger than
+    /// 16 bytes.
+    pub fn try_declare_var(
+        &mut self,
+        var: Variable,
+        ty: Type,
+        stack_map_behavior: StackMapBehavior,
+    ) -> Result<(), DeclareVariableError> {
+        if stack_map_behavior == StackMapBehavior::Include {
+            assert!(
+                ty.bytes() <= 16,
+                "types larger than 16 bytes cannot be in stack maps"
+            );
+        }
         if self.func_ctx.types[var] != types::INVALID {
             return Err(DeclareVariableError::DeclaredMultipleTimes(var));
         }
         self.func_ctx.types[var] = ty;
+        if stack_map_behavior == StackMapBehavior::Include {
+            self.func_ctx.stack_map_vars.insert(var);
+        }
         Ok(())
     }
 
-    /// In order to use a variable (by calling [`FunctionBuilder::use_var`]), you need
-    /// to first declare its type with this method.
-    pub fn declare_var(&mut self, var: Variable, ty: Type) {
-        self.try_declare_var(var, ty)
+    /// Declares the type of a variable, panicking if it is already declared.
+    ///
+    /// # Panics
+    ///
+    /// Panics if either the variable has already been declared, or if
+    /// `stack_map_behavior` is `StackMapBehavior::Include` and the given type
+    /// is larger than 16 bytes.
+    pub fn declare_var(&mut self, var: Variable, ty: Type, stack_map_behavior: StackMapBehavior) {
```

Comment:
Would it be possible to introduce a new method instead. Most users of cranelift-frontend will not use stackmaps and may not even know what they are and thus not know what they should pass.

Reply:
I actually went back and forth on this, but after switching to a different method to declare that a variable needs inclusion in stack maps, rather than a parameter to `declare_var` (see most recent commit) I actually like the different method better. Thanks for the nudge!